iModelCore/ecobjects/src/ECSchema.cpp

```diff
@@ -3216,11 +3216,25 @@ ECSchemaPtr SearchPathSchemaFileLocater::_LocateSchema(SchemaKeyR key, SchemaMat
 +---------------+---------------+---------------+---------------+---------------+------*/
 ECSchemaPtr StringSchemaLocater::_LocateSchema(SchemaKeyR key, SchemaMatchType matchType, ECSchemaReadContextR schemaContext)
     {
-    ECSchemaPtr schema;
-    auto schemaXml = m_schemaXml.find(key.m_schemaName);
-    if (schemaXml != m_schemaXml.end())
-        ECSchema::ReadFromXmlString(schema, schemaXml->second.c_str(), schemaContext);
-    return schema;
+    bmap<Utf8String, std::tuple<Utf8String, SchemaKey, ECSchemaPtr>>::iterator iter = m_knownSchemas.find(key.GetName());
+    if (iter == m_knownSchemas.end())
+        return nullptr;
+
+    const std::tuple<Utf8String, SchemaKey, ECSchemaPtr>& knownSchema = iter->second;
+    Utf8StringCR expectedSchemaXml =std::get<0>(knownSchema);
+    SchemaKeyCR expectedSchemaKey = std::get<1>(knownSchema);
+    ECSchemaPtr expectedECSchema = std::get<2>(knownSchema);
+    if (!expectedSchemaKey.Matches(key, matchType))
+        return nullptr;
+    if (expectedECSchema.IsValid())
+        return expectedECSchema;
+
+    // Read schema from Xml string if the cached schema is invalid
+    ECSchemaPtr schemaOut;
+    if (SchemaReadStatus::Success != ECSchema::ReadFromXmlString(schemaOut, expectedSchemaXml.c_str(), schemaContext))
+        return nullptr;
+    m_knownSchemas[key.GetName()] = std::make_tuple(expectedSchemaXml, expectedSchemaKey, schemaOut);
+    return schemaOut;

```

Comment:
Schema locaters do not need to cache the schemas it loads, the SchemaContext the locater is added to does the caching.  So this doesn't need to be so complex